presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/ExtendedHiveMetastore.java

```diff
@@ -77,7 +77,7 @@
 
     MetastoreOperationResult dropColumn(MetastoreContext metastoreContext, String databaseName, String tableName, String columnName);
 
-    Optional<Partition> getPartition(MetastoreContext metastoreContext, String databaseName, String tableName, List<String> partitionValues);
+    Optional<Partition> getPartition(MetastoreContext metastoreContext, Table table, List<String> partitionValues);
 
     Optional<List<String>> getPartitionNames(MetastoreContext metastoreContext, String databaseName, String tableName);
```

Comment:
Since we are introducing table as an argument to other relevant methods, I think we can also make a similar change for `getPartitionNames`. This will help in reducing multiple `getTable` calls when called from `SemiTransactionalHiveMetastore` since we already fetch the table object there.

Reply:
This is a valid ask. However, MockHiveMetastoreClient gets called from TestCachingHiveMetastore tests internally and this client is implemented in a way that it checks for BAD_DATABASE for all get calls. For example, `testInvalidGetPartitionNames` in TestCachingHiveMetastore will ultimately call `getPartitionNames` in MockHiveMetastoreClient which verifies database name. 

Having said that, I have a plan in my mind. To incorporate the optimisation that you proposed, we need to remove db name check from getTable() method in MockHiveMetastoreClient (https://github.com/prestodb/presto/blob/1c9de99de7cb65b97f0bb304c3e58fb4caf02dac/presto-hive-metastore/src/test/java/com/facebook/presto/hive/metastore/thrift/MockHiveMetastoreClient.java#L165) and remove the corresponding test case from TestCachingHiveMetastore (https://github.com/prestodb/presto/blob/1c9de99de7cb65b97f0bb304c3e58fb4caf02dac/presto-hive/src/test/java/com/facebook/presto/hive/metastore/TestCachingHiveMetastore.java#L153). This essentially means MockHiveMetastoreClient will always return a table object irrespective of the database name passed. But this is only a test case and should not cause any issues as such. Once this is done, we can do the changes that you proposed @imjalpreet .

Good to get some committers eyes on this as well. @tdcmeehan @prestodb/committers @NikhilCollooru @highker .

Reply:
@prestodb/committers I pushed the second commit separately to implement what I proposed in the above comment. Feel free to have a look and share your thoughts on this.